bokehjs/test/defaults/index.ts

```diff
@@ -28,63 +26,36 @@ function get_defaults(name: string) {
     throw new Error(`can't find defaults for ${name}`)
 }
 
-function safe_stringify(v: unknown): string {
-  if (v === Infinity) {
-    return "Infinity"
-  } else {
-    try {
-      return JSON.stringify(v)
-    } catch {
-      return `${v}`
-    }
-  }
-}
-
-function deep_value_to_serializable(_key: string, value: unknown, _optional_parent_object: any): any {
-  if (value instanceof HasProps) {
-    return {type: value.type, attributes: value.attributes_as_json()}
-  } else if (isArray(value)) {
-    const ref_array: any[] = []
-    for (let i = 0; i < value.length; i++) {
-      ref_array.push(deep_value_to_serializable(i.toString(), value[i], value))
-    }
-    return ref_array
-  } else if (isPlainObject(value)) {
-    const ref_obj: {[key: string]: any} = {}
-    for (const [subkey, subvalue] of entries(value)) {
-      ref_obj[subkey] = deep_value_to_serializable(subkey, subvalue, value)
-    }
-    return ref_obj
-  } else
-    return value
-}
-
 type KV = {[key: string]: any}
 
-function check_matching_defaults(name: string, python_defaults: KV, bokehjs_defaults: KV): boolean {
+function check_matching_defaults(name: string, python_defaults: KV, bokehjs_defaults: KV, serializer: Serializer): boolean {
   const different: string[] = []
   const python_missing: string[] = []
   const bokehjs_missing: string[] = []
 
-  for (const [k, js_v] of entries(bokehjs_defaults)) {
-    // special case for graph renderer node_renderer
-    if (name === "GraphRenderer" && k === "node_renderer")
+  for (let [k, js_v] of entries(bokehjs_defaults)) {
+    // special case for graph renderer node_renderer/edge_renderer
+    if (name == "GraphRenderer" && (k == "node_renderer" || k == "edge_renderer"))
       continue
 
-    // special case for graph renderer node_renderer
-    if (name === "GraphRenderer" && k === "edge_renderer")
+    // special case for factor range and start/end (null vs 0)
+    if (name == "FactorRange" && (k == "start" || k == "end"))
       continue
 
     // special case for date picker, default is "now"
-    if (name === 'DatePicker' && k === 'value')
+    if (name == 'DatePicker' && k == 'value')
       continue
 
     // special case for date time tickers, class hierarchy and attributes are handled differently
-    if (name === "DatetimeTicker" && k === "tickers")
+    if (name == "DatetimeTicker" && k == "tickers")
+      continue
+
+    // special case for days/months/years tickers (null vs computed integer)
+    if ((name == "DaysTicker" || name == "MonthsTicker" || name == "YearsTicker") && k == "interval")
       continue
 
     // special case for Title derived text properties
-    if (name === "Title" && (k === "text_align" || k === "text_baseline"))
+    if (name == "Title" && (k == "text_align" || k == "text_baseline"))
```

Comment:
FWIW not 100% sure all these special cases above still apply

Reply:
I will need to give defaults's tests another iteration anyway, because although this PR improves things a lot, the analysis is still weak and there are too many special cases.